src/Components/test/E2ETest/ServerExecutionTests/ServerSideAppTest.cs

```diff
@@ -96,7 +95,7 @@ public void HasFetchDataPage()
         {
             // Navigate to "Fetch Data"
             Browser.FindElement(By.LinkText("Fetch data")).Click();
-            Browser.Equal("Weather forecast", () => Browser.FindElement(By.TagName("h1")).Text);
+            Browser.Equal("Weather forecast", () => Browser.FindElement(By.CssSelector("h1#fetch-data")).Text);
```

Comment:
This can happen because we click on the link above and the navigation happens between us finding the current h1 and the one in the page. (Produces a stale reference exception). Use a more specific CSS selector to prevent it.

Reply:
> Use a more specific CSS selector to prevent it.

I don't think this is correct. The behavior of `WaitAssertCore` is to keep trying the assertion until it passes (until some timeout). Using a more specific selector won't change this, because it would have kept retrying anyway if the assertion failed. Or am I missing something?

I'm not saying your change is wrong; I'm just saying that there must be some other cause for any flakiness here.

Reply:
The issue was a stale element exception.
What likely happened was that the page changed from Index to FetchData while Browser.FindElement was executing, so it would find H1 on index and then that node would become detached from the DOM when accessing text, which would trigger the exception.

By using the more specific selector, the first find would fail until we navigated to the new page and find the h1 we were looking for. In general, based on my experience, it pays off to be more specific here.

That's only my theory of what was happening here.

Reply:
The waitassert logic is here: https://github.com/aspnet/AspNetCore/blob/9f1a978230cdd161998815c425bfd2d25e8436b6/src/Shared/E2ETesting/WaitAssert.cs#L55

It does a `try`/`catch` around the assertion code, so if it throws for any reason, we just keep retrying in case it passes before the eventual timeout. As such I don't see how it would make any difference whether the first find fails or not.

Reply:
Ok. My bad this is actually a timeout that is hidden by this line:
https://github.com/aspnet/AspNetCore/blob/master/src/Shared/E2ETesting/WaitAssert.cs#L69

There is no way to know the real cause I believe as re-running the assertion is not correct.
Given this is the case, we should modify WaitAssertCore to capture the last exception in a local and rethrow that.

Also, how often does this thing poll?

Reply:
> we should modify WaitAssertCore to capture the last exception in a local and rethrow that.

Isn't that the same as what it already does? It already keeps running the same callback over and over, and if the `WebDriverWait` times out, then it runs the same callback one more time but this time lets any exception bubble up the call stack.

> Also, how often does this thing poll?

I'm not certain if our code modifies the defaults anywhere, but apparently the default is to poll every 500ms (https://stackoverflow.com/a/44645991).

Reply:
> Isn't that the same as what it already does?

No. It reruns the assertion and that can cause a different exception than the last one.

Reply:
It's runs the assertion N times in a loop until there's a timeout, then runs it again for the (N+1)th time. For what reason would we prefer to report the Nth exception instead of the (N+1)th exception?

Reply:
I believe so. If not, I can't think how that test failed.